### PR TITLE
[Intel GPU] Dispatch Stub support

### PR DESCRIPTION
**Pull request by ZhiweiYan-96**
# Motivation
Structured codegen is beneficial for easier decoupling tensor meta setting and kernel implementation. At present, XPU operators need to handle tensor metas in hand-written way. 

We plan to leverage the codegen system for auto generate structured operators. This PR facilitate the `DispatchStub` support for  Intel GPUs. Based on that, XPU operators would have possibility to register kernel functor to operator stubs.

This is a prerequisite of PR #130082, where we will modify the codegen system to generate XPU needed source files and headers.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130088
* #130082
* __->__ #130019

cc @gujinghui @EikanWang @fengyuan14 @guangyey

### CODE REVIEW

**Review by EikanWang — APPROVED** [review]

**Review by gujinghui — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -1173,6 +1173,10 @@ if(APPLE)
   append_cxx_flag_if_supported("-Wno-missing-braces" CMAKE_CXX_FLAGS)
 endif()
 
+if(USE_XPU)
```
Comment (albanD) [inline CMakeLists.txt:1176]: Do we expect any platform limitation being needed here?

File: aten/src/ATen/native/DispatchStub.h [hunk 2]
```diff
@@ -183,6 +184,7 @@ struct TORCH_API DispatchStubImpl {
     void* cuda_dispatch_ptr = nullptr;
     void* hip_dispatch_ptr = nullptr;
     void* mps_dispatch_ptr = nullptr;
+    void* xpu_dispatch_ptr = nullptr;
```
Comment (albanD) [inline aten/src/ATen/native/DispatchStub.h:187]: For struct size could we also guard this with USE_XPU ?
We can also guard set_xpu_dispatch_ptr with this and make it either an error if called outside of an XPU build

**Review by albanD — COMMENTED** [review]
Sounds good. Only small improvements.

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by ZhiweiYan-96** [comment]
@albanD  Thank for your approval!

**Comment by ZhiweiYan-96** [comment]
Meet new ci failure, I am fixing

**Review by EikanWang — APPROVED** [review]

**Comment by ZhiweiYan-96** [comment]
hi @albanD @EikanWang 

The new failure in CI is caused by the fact that  `CPPExtension`  related UTs compile source files without `-DUSE_XPU` while it links to `libtorch` build with such build variable in XPU ci flow. This results in the runtime failure.  We verify this by adding extra_cflags=["-g", "-DUSE_XPU" in uts and it passed.

Considering that, XPU does not support cppextension at present, would it acceptable to your that we remove the `USE_XPU` marco in source file and let it default exists, just like cuda/mps?

**Comment by EikanWang** [comment]
I think it is an alternative to skip CPP_Eextension test cases as Intel GPU has not supported it well.  @albanD , may I know which approach is your prefer?

Approach 1: Skip CPP_Extension test cases for Intel GPU
Approach 2: Remove the macro block and add it back when the CPP_Extension is available for Intel GPU.

Thanks,
Eikan

**Comment by albanD** [comment]
Skipping cpp_extensions test is definitely an option but that is a widely used features by third party repos. So you most likely want to fix that sooner rather than later.
I'm not sure to understand what is being done differently in USE_XPU flag handling compared to other backends that lead to this though.

**Comment by EikanWang** [comment]
> Skipping cpp_extensions test is definitely an option but that is a widely used features by third party repos. So you most likely want to fix that sooner rather than later. 

Yes, we plan to support cpp_extension and a community contributor (not from Intel) jas just submitted a [PR](https://github.com/pytorch/pytorch/pull/131276) to support the feature. But we still have some a open for the solution. we need some time to conclude the open.

> I'm not sure to understand what is being done differently in USE_XPU flag handling compared to other backends that lead to this though.

For XPU CI, `USE_XPU` is `True` to build all torch libraries. But it is `FALSE` in `cpp_extesnion` because we do not set it explicitly for XPU. We need to keep the `USE_XPU` being consistent just like AMD.

https://github.com/pytorch/pytorch/blob/main/torch/utils/cpp_extension.py#L235

**Comment by EikanWang** [comment]
> > For XPU CI, USE_XPU is True to build all torch libraries. But it is FALSE in cpp_extesnion because we do not set it explicitly for XPU.
> 
> Why not? Don't we re-use the torch compilation flag during cpp_extension build usually? I would expect that's how we transfer cuda/rocm/mps related flags?

Yes, we should re-use the torch compilation flag. And we did it in another PR(https://github.com/pytorch/pytorch/pull/131276) as `cpp_extension` is an independent feature for XPU.

**Comment by EikanWang** [comment]
@ZhiweiYan-96 , please help refine the PR a little bit.

**Comment by ZhiweiYan-96** [comment]
The failed cppextension uts has been skipped, thanks for comments :)

**Pull request closed without merging**